Turn dopple into a proper Python package

cburgdorf opened the pull request:
## What's wrong

The repo is currently unmaintained. To adopt it under the snake charmers umbrella, we need to do bunch of things.

## How was it fixed.

Not yet fixed but this PR does these two things:

- [x] Add basic ethereum-python-template-project scafolding
- [x] Bring it on [pypi](https://pypi.org/project/dopple/0.4.0a0/)

cburgdorf commented:
Hey @axic thank you for the feedback!

>license change

I moved this back so that we stick to Apache 2.0 for the time being. 

>readme overhaul

I put the Readme changes in a separate commit now

>adding __init__.py
>duplicating dopple.py as dopple/dopple.py

I accidentally left the old `dopple.py` in the root. (Sorry, the PR was pretty much a WIP, I wasn't expecting everyone to jump on it :sweat_smile:)

I removed it now so that the one and only `dopple.py` is under the `dopple` directory. Also adding the `__init__.py` is part of the required things to make this a real python module.

There's one important thing though: When this lands `dopple` would need to be installed using `pip`.

I'm pretty sure we can add other distribution mechanisms like distribution through a binary file (maybe even trying out [pyoxidizer](https://pyoxidizer.readthedocs.io/en/latest/)) but I want to make sure we are not ruining anybody's setup here.

As far as I can tell aleth uses this repo as a submodule but it's pinned to a specific commit so it shouldn't be immediately affected (cc @chfast)

I'd ideally like to get this in because it sets up testing, CI, linting, release notes etc. After getting CI running and doing some other chores I can look into how to create binaries.

chfast commented on setup.py:
```diff
@@ -0,0 +1,82 @@
+    url='https://github.com/ethereum/dopple',
+    include_package_data=True,
+    install_requires=[
+        "eth-utils>=1,<2",
```
Why?

cburgdorf replied:
Oh good catch, I guess this one is in the defaults of https://github.com/ethereum/ethereum-python-project-template because it is widely used but I agree that it should be removed in this case.

cburgdorf commented:
>This makes it unusable without "install", right?

Yes, because it tries to read the version from the package via `pkg_resources.get_distribution("dopple").version` which won't work if the package isn't installed.

But actually, maybe we can create a fallback for that and keep it runnable as a pure script at least until we have figured out how to also distribute it as an executable.

chfast commented:
> But actually, maybe we can create a fallback for that and keep it runnable as a pure script at least until we have figured out how to also distribute it as an executable.

That would nice, but I expect what will not be supported forever as you start adding external dependencies. On the other hand, when this is published, it might be easier to ask users to do `pip install dopple; dopple aleth ...` instead of `scripts/dopple/dopple.py aleth ...`.

cburgdorf commented:
>That would nice

I added this [fallback](https://github.com/ethereum/dopple/pull/4/files#diff-72b2158bbe4fa5cb5330bbb9a3b683f9R36) and also added a paragraph to the README that explains it can be run as a pure script. It will show up as `unknown` version when just invoked as a script but I hope that's an ok tradeoff for now :)

>but I expect what will not be supported forever as you start adding external dependencies

Yes, but maybe we can keep the non-installation support around as long we don't support binary distribution yet. We can also add a job to CI to ensure we won't accidentally break it.

Any objections to get this merged?

cburgdorf commented:
@chfast @axic are you ok with it being merged in its current form?

chfast commented on .circleci/config.yml:
```diff
@@ -0,0 +1,77 @@
+version: 2.0
```
Do you even run this on circle CI?

cburgdorf replied:
That's the plan. I want to setup CI jobs so that we can run some tests and ensure we won't break things while we apply some refactoring.

chfast commented on .circleci/config.yml:
```diff
@@ -0,0 +1,77 @@
+
+jobs:
+  docs:
+    <<: *common
```
You don't need to use YAML references in Circle CI 2.1

chfast commented on .project-template/fill_template_vars.sh:
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
```
Is this script needed?

cburgdorf replied:
It's handy, if we ever decide to change the name or the description etc, it will go through all places that needs to be updated. It's all based on our default scaffolding that we use for all projects https://github.com/ethereum/ethereum-python-project-template

cburgdorf commented on .circleci/merge_pr.sh:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
As far as I understand, we run CI jobs against PRs that are merged in the base. That is to prevent CI runs go green when in reality they create problems after they are merged. So it's another safety measure.

cburgdorf commented on README.md:
````diff
@@ -1,20 +1,109 @@
+## Installation
+
+```sh
+pip install dopple
````
Absolutely.

cburgdorf commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,21 @@
+[//]: # (See: https://dopple.readthedocs.io/en/latest/contributing.html#pull-requests)
+- [ ] Add entry to the [release notes](https://github.com/ethereum/dopple/blob/master/newsfragments/README.md)
+
+#### Cute Animal Picture
```
What about grumpy cat pictures?

![](https://newscdn.weigelbroadcasting.com/Xx052-1558095817-tile_image-136842.jpg)

joking, no problem, not a hard requirement, just the defaults of https://github.com/ethereum/ethereum-python-project-template

pipermerriam commented:
> I'm not convinced that adding a ton of scripts that are not currently used is great idea.

Thanks for your patience.  I know it's a lot of change.  The project template is a way for us (the snake charmers team) to have the tools we're accustomed to available consistently across the repos we maintain.  Everything in here is pretty standard for us and going forward we'll plan on taking responsibility for this library now that it's going to be a dependency of Trinity.
